### DISCUSSION TITLE
Loader FrozenImporter cannot handle module esptool

### DISCUSSION THREAD

**Discussion by fab-uleuh** [pyinstaller/pyinstaller#7120] [Help] [answered]
Hello,

I try to use pyinstaller with my script.

pyinstaller --hiddenimport=esptool -F main.py

I've the error : 
Loader FrozenImporter cannot handle module esptool

In debug:

maximum recursion depth exceeded while calling a Python object
Please install esptool
\# destroy esptool
\# destroy esptool
\# destroy esptool

python3.8 (I tried with 3.6 also)
pyinstaller==5.4.1

Thanks

**Comment by bwoodsend** [comment, marked as answer]
Ughh, That's because `esptool` rather foolishly named their entry-point script `esptool.py`. When you run `pyinstaller [any args here]`, Python adds the location of the `pyinstaller` script (usually something like `/usr/lib/python3.8/bin`) to `sys.path`. That directory is also where the `esptool.py` command line script lives so with it added to `sys.path`, `import esptool` imports `esptool.py` the tool rather than `esptool` the library. Now `esptool.py` the tool contains `import esptool` which, due to this `sys.path` shenanigans will import itself rather than the library `esptool` which is where this infinite recursion comes from.

You can get around it by using `python -m PyInstaller` instead of just `pyinstaller`.

**Reply by fab-uleuh** [reply]
Thank you @bwoodsend  it's work! 

Did you know how to work around with fbs ?

**Reply by bwoodsend** [reply]
What's _fbs_?

**Reply by fab-uleuh** [reply]
https://github.com/mherrmann/fbs-tutorial

**Reply by fab-uleuh** [reply]
They use pyinstaller to package

**Reply by rokm** [reply]
If you cannot influence how PyInstaller is invoked, you could also remove/rename those `esptool.py`, `espefuse.py`, and `espsecure.py` scripts from the scripts directory.

**Reply by bwoodsend** [reply]
Does `python -m fbs` not do the job?

**Reply by fab-uleuh** [reply]
Unfortunately no

**Reply by bwoodsend** [reply]
Ahh, forget that. `fbs` [makes the classic mistake of expecting a Python installation's `/bin`/`/Scipts` directory to be first in `$PATH`](https://github.com/mherrmann/fbs/blob/b5777fb20dfd3a397a4f1740f2fdc1946dad6d42/fbs/freeze/__init__.py#L22-L28).

**Reply by bwoodsend** [reply]
@rokm We could make [this](https://github.com/mherrmann/fbs/blob/b5777fb20dfd3a397a4f1740f2fdc1946dad6d42/fbs/freeze/__init__.py#L22-L28) point to a wrapper function which looks like:

```python
def _console_script_run():
     sys.path.pop(0)
     run()
```
There's no reason for the contents of an entry-points `bin` directory to be importable?

**Reply by fab-uleuh** [reply]
I make modification on __init_.py and it's work 

    args = [
        'python',
        '-m',
        'PyInstaller',

Thank you !

**Reply by rokm** [reply]
> @rokm We could make [this](https://github.com/mherrmann/fbs/blob/b5777fb20dfd3a397a4f1740f2fdc1946dad6d42/fbs/freeze/__init__.py#L22-L28) point to a wrapper function which looks like:
> 
> ```python
> def _console_script_run():
>      sys.path.pop(0)
>      run()
> ```
> 
> There's no reason for the contents of an entry-points `bin` directory to be importable?

Indeed. Removing scripts directory from `sys.path` sounds like a good idea.
